hazelcast/src/main/java/com/hazelcast/internal/partition/operation/AbstractPartitionOperation.java

```diff
@@ -67,16 +68,51 @@
         return operations;
     }
 
-    final Operation createFragmentReplicationOperation(PartitionReplicationEvent event, ServiceNamespace ns) {
+    final Operation createFragmentReplicationOperation(PartitionReplicationEvent event, ServiceNamespace ns,
+            String serviceName) {
         NodeEngineImpl nodeEngine = (NodeEngineImpl) getNodeEngine();
-        FragmentedMigrationAwareService service = nodeEngine.getService(ns.getServiceName());
+        FragmentedMigrationAwareService service = nodeEngine.getService(serviceName);
+
+        assert service.isKnownServiceNamespace(ns) : ns + " should be known by " + service;
         Operation op = service.prepareReplicationOperation(event, singleton(ns));
         if (op != null) {
-            op.setServiceName(ns.getServiceName());
+            op.setServiceName(serviceName);
         }
         return op;
     }
 
+    final Collection<Operation> createFragmentReplicationOperations(PartitionReplicationEvent event, ServiceNamespace ns) {
+        Collection<Operation> operations = emptySet();
+        NodeEngineImpl nodeEngine = (NodeEngineImpl) getNodeEngine();
+        Collection<ServiceInfo> services = nodeEngine.getServiceInfos(FragmentedMigrationAwareService.class);
+
+        for (ServiceInfo serviceInfo : services) {
+            FragmentedMigrationAwareService service = serviceInfo.getService();
+            if (!service.isKnownServiceNamespace(ns)) {
+                continue;
+            }
+
+            Operation op = service.prepareReplicationOperation(event, singleton(ns));
+            if (op == null) {
+                continue;
+            }
+
+            op.setServiceName(serviceInfo.getName());
+
+            if (operations.isEmpty()) {
+                // generally a namespace belongs to a single service only
+                operations = singleton(op);
```

Comment:
My reasoning is, to avoid `ArrayList` creating, if a namespace belongs to a single service. Currently only lock service shares namespaces with map and multimap.